File: src/main/java/org/taktik/icure/logic/impl/filter/service/ServiceByContactsAndSubcontactsFilter.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl.filter.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.entities.Contact;
import org.taktik.icure.entities.embed.Service;
import org.taktik.icure.entities.embed.ServiceLink;
import org.taktik.icure.entities.embed.SubContact;
import org.taktik.icure.logic.ContactLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.SessionLogic;
import org.taktik.icure.logic.impl.filter.Filter;
import org.taktik.icure.logic.impl.filter.Filters;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;


@org.springframework.stereotype.Service
public class ServiceByContactsAndSubcontactsFilter implements Filter<String, Service, org.taktik.icure.dto.filter.service.ServiceByContactsAndSubcontactsFilter> {
	ContactLogic contactLogic;
	ICureSessionLogic sessionLogic;

	@Autowired
	public void setContactLogic(ContactLogic contactLogic) {
		this.contactLogic = contactLogic;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	public Set<String> resolve(org.taktik.icure.dto.filter.service.ServiceByContactsAndSubcontactsFilter filter, Filters context) {
        Set<String> ids = new HashSet<>();
        for (Contact c : contactLogic.getContacts(filter.getContacts())) {
            if (filter.getSubContacts()!=null) {
                for (SubContact sc: c.getSubContacts()) {
                    if (filter.getSubContacts().contains(sc.getId()) && sc.getServices() != null) {
                        ids.addAll(sc.getServices().stream().map(ServiceLink::getServiceId).collect(Collectors.toList()));
                    }
                }
            } else {
                ids.addAll(c.getServices().stream().map(Service::getId).collect(Collectors.toList()));
            }
         }

        return ids;
    }
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/filter/service/ServiceByContactsAndSubcontactsFilter.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl.filter.service

import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.flatMapConcat
import org.springframework.stereotype.Service
import org.taktik.icure.asynclogic.ContactLogic
import org.taktik.icure.asynclogic.impl.filter.Filter
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.domain.filter.service.ServiceByContactsAndSubcontactsFilter

@FlowPreview
@Service
class ServiceByContactsAndSubcontactsFilter(private val contactLogic: ContactLogic) : Filter<String, org.taktik.icure.entities.embed.Service, ServiceByContactsAndSubcontactsFilter> {

    override fun resolve(filter: ServiceByContactsAndSubcontactsFilter, context: Filters): Flow<String> {
        val contacts = contactLogic.getContacts(filter.contacts)
        return if (filter.subContacts != null) {
            contacts.flatMapConcat { c -> c.subContacts.flatMap { sc -> if (filter.subContacts!!.contains(sc.id)) sc.services.mapNotNull { it.serviceId } else listOf() }.asFlow() }
        } else {
            contacts.flatMapConcat { c -> c.services.map { it.id }.asFlow() }
        }
    }
}
